proxy/proxy.go

```diff
@@ -315,134 +279,54 @@ func tryCatch(p func(), onErr func(err interface{})) {
-		Host:             r.Host,
-		RemoteAddr:       r.RemoteAddr,
-		RequestURI:       r.RequestURI,
-		TLS:              r.TLS}
-}
-
-func cloneResponseMetadata(r *http.Response) *http.Response {
-	return &http.Response{
-		Status:           r.Status,
-		StatusCode:       r.StatusCode,
-		Proto:            r.Proto,
-		ProtoMajor:       r.ProtoMajor,
-		ProtoMinor:       r.ProtoMinor,
-		Header:           cloneHeader(r.Header),
-		Body:             &bodyBuffer{&bytes.Buffer{}},
-		ContentLength:    r.ContentLength,
-		TransferEncoding: r.TransferEncoding,
-		Close:            r.Close,
-		Request:          r.Request,
-		TLS:              r.TLS}
-}
-
-func (c *filterContext) ResponseWriter() http.ResponseWriter { return c.w }
-func (c *filterContext) Request() *http.Request              { return c.req }
-func (c *filterContext) Response() *http.Response            { return c.res }
-func (c *filterContext) MarkServed()                         { c.served = true }
-func (c *filterContext) Served() bool                        { return c.served || c.servedWithResponse }
-func (c *filterContext) PathParam(key string) string         { return c.pathParams[key] }
-func (c *filterContext) StateBag() map[string]interface{}    { return c.stateBag }
-func (c *filterContext) BackendUrl() string                  { return c.backendUrl }
-func (c *filterContext) OriginalRequest() *http.Request      { return c.originalRequest }
-func (c *filterContext) OriginalResponse() *http.Response    { return c.originalResponse }
-func (c *filterContext) OutgoingHost() string                { return c.outgoingHost }
-func (c *filterContext) SetOutgoingHost(h string)            { c.outgoingHost = h }
-
-func (c *filterContext) Serve(res *http.Response) {
-	res.Request = c.Request()
-
-	if res.Header == nil {
-		res.Header = make(http.Header)
-	}
-
-	if res.Body == nil {
-		res.Body = &bodyBuffer{&bytes.Buffer{}}
-	}
-
-	c.servedWithResponse = true
-	c.res = res
-}
-
-// creates an empty shunt response with the initial status code of 404
-func shunt(r *http.Request) *http.Response {
-	return &http.Response{
-		StatusCode: http.StatusNotFound,
-		Header:     make(http.Header),
-		Body:       &bodyBuffer{&bytes.Buffer{}},
-		Request:    r}
-}
-
 // applies all filters to a request
-func (p *Proxy) applyFiltersToRequest(f []*routing.RouteFilter, ctx *filterContext, onErr func(err interface{})) []*routing.RouteFilter {
-	var start time.Time
+func (p *Proxy) applyFiltersToRequest(f []*routing.RouteFilter, ctx *context) []*routing.RouteFilter {
+	filtersStart := time.Now()
+
 	var filters = make([]*routing.RouteFilter, 0, len(f))
 	for _, fi := range f {
-		start = time.Now()
-		tryCatch(func() { fi.Request(ctx) }, onErr)
+		start := time.Now()
+		tryCatch(func() { fi.Request(ctx) }, func(err interface{}) {
+			if p.flags.Debug() {
+				ctx.debugFilterPanics = append(ctx.debugFilterPanics, err)
+				return
+			}
+
+			log.Error("error while processing filters during request: ", err)
```

Comment:
this is due to one of our design debts: currently the filter extension interface doesn't tell which filter this is. I would better postpone fixing this issue instead of using e.g. an index. Or shall we print the index for now?